text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,523 @@
+
+* This is done during stack synthesis.
+* Recursion stops upon encountering a contained `Stack`, `Stage` or `NestedStack`.
+* Contrary to today, nested stacks will each have their own copy of a metadata resource.
+
+## Obtaining construct identifiers
+
+From a construct instance, we need to obtain a construct identifier. To not
+proliferate variants of identifiers too much, it's preferable for this identifier
+to be the same, or substantially the same, as the class' jsii type name.
+
+We are planning to eventually introduce a feature for jsii-powered Run-Time
+Type Information (RTTI). We will use this project to start introducing that
+feature at its very simplest, and the initial feature will be to retrieve a jsii
+FQN and module version from an object instance. See Appendix A for a description
+of how jsii RTTI will work.
+
+For the initial version of Metadata v2, we will simply say that the construct
+identifier is whatever jsii RTTI returns.
+
+> In the future, we may want to onboard libraries that are not jsii-libraries (either
+> written in plain TypeScript or a jsii client language like Java); if that happens,
+> we will add a CDK-specific core library function to attach a specific construct
+> identifier to a construct instance, like
+> `CdkMetadata.describe(construct, "@my/library.MyConstruct@1.2.3")`.
+
+## Filtering down the list of construct identifiers
+
+To protect the privacy of our users, we only want to report on 1st-party
+constructs, authored and vended by AWS.
+
+We decided against an open system, instead preferring to have a specific
+allow-list of packages (or package prefixes) we are going to report on. For
+example, the current list looks like this and can be found
+[here](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/core/lib/private/runtime-info.ts):
+
+```ts
+const ALLOWLIST_SCOPES = ['@aws-cdk', '@aws-solutions-konstruk', '@aws-solutions-constructs', '@amzn'];
+const ALLOWLIST_PACKAGES = ['aws-rfdk'];
+```
+
+Having a PR opened against the CDK to update this list will be a clear
+interaction moment between an AWS team trying to onboard with this mechanism
+and us, and will prevent 3rd party authors from accidentally (or maliciously)
+onboarding themselves as well and sending us information that we don't have
+a use-case for.
+
+## Metadata resource encoding
+
+Once we have the list of construct identifiers, we need to encode it into the
+metadata resource.
+
+The Metadata resource looks like this:
+
+```yaml
+Resources:
+  CDKMetadata:
+    Type: AWS::CDK::Metadata
+    Properties:
+      Modules: <v1 module information, deprecated by this RFC>
+      Analytics: <v2 construct information, introduced by this RFC>
+```
+
+### Construct list size
+
+The list of construct identifiers we have discovered may be sizeable. A rough
+estimate of the maximum number of constructs used in a stack will be about
+400 (maximum of 200 L1s with a corresponding L2 wrapper). We will therefore
```

Comment:
Does it make sense to record both L2 and L1 in case the L2 is used? It actually makes sense to omit the L1 in this case I think so we can easily identify direct L1 usage (which is of particular interest for our construct library planning).

The simplest solution would be to omit L1s that are named `Resource` and that their patent construct inherits from the Resource base class.

Reply:
I've been on the fence about this, and trying to come up with a more general rule that would also cover this (but failing to).

But I guess it makes sense to just special-case it, you're right.